lib/libuzfs/libuzfs_attr.c

```diff
@@ -0,0 +1,590 @@
+}
+
+static ssize_t
+libuzfs_inode_get_kvattr_zap(sa_handle_t *sa_hdl, sa_attr_type_t *sa_tbl,
+    const char *name, char *value, size_t size)
+{
+	// get zap object from sa
+	objset_t *os = sa_hdl->sa_os;
+	uint64_t xattr_zap_obj;
+	int err = sa_lookup(sa_hdl, sa_tbl[UZFS_ZXATTR],
+	    &xattr_zap_obj, sizeof (uint64_t));
+	if (err != 0) {
+		return (-err);
+	}
+
+	uint64_t integer_size = 0;
+	uint64_t integer_num = 0;
+	err = zap_length(os, xattr_zap_obj, name,
+	    &integer_size, &integer_num);
+	if (err != 0) {
+		return (-err);
+	}
+
+	ASSERT(integer_size == 1);
+	ASSERT(integer_num <= size);
+	err = zap_lookup(os, xattr_zap_obj, name, 1, size, value);
+	if (err != 0) {
+		return (-err);
+	}
+
+	return (integer_num);
+}
+
+static ssize_t
+libuzfs_inode_get_kvattr_sa(sa_handle_t *sa_hdl, const sa_attr_type_t *sa_tbl,
+    const char *name, char *value, size_t size)
+{
+	nvlist_t *nvl;
+	int err = libuzfs_get_nvlist_from_handle(sa_tbl, &nvl, sa_hdl);
+	if (err != 0) {
+		return (-err);
+	}
+
+	uchar_t *nv_value;
+	uint_t nv_size = 0;
+	err = nvlist_lookup_byte_array(nvl, name, &nv_value, &nv_size);
+	if (err == 0 && nv_size <= size) {
+		memcpy(value, nv_value, nv_size);
+	}
+	nvlist_free(nvl);
+	if (err != 0) {
+		return (-err);
+	}
+
+	if (nv_size > size) {
+		return (-ERANGE);
+	}
+	return (nv_size);
+}
+
+ssize_t
+libuzfs_inode_get_kvattr(libuzfs_dataset_handle_t *dhp, uint64_t ino,
+    const char *name, char *value, uint64_t size, int flags)
+{
+	sa_handle_t *sa_hdl = NULL;
+	int err = sa_handle_get(dhp->os, ino, NULL, SA_HDL_PRIVATE, &sa_hdl);
+	if (err != 0) {
+		return (-err);
+	}
+
+	ssize_t rc = libuzfs_inode_get_kvattr_sa(sa_hdl, dhp->uzfs_attr_table,
+	    name, value, size);
+	if (rc == -ENOENT) {
+		rc = libuzfs_inode_get_kvattr_zap(sa_hdl, dhp->uzfs_attr_table,
+		    name, value, size);
+	}
+
+	sa_handle_destroy(sa_hdl);
+	return (rc);
+}
+
+static int
+libuzfs_inode_remove_kvattr_from_sa(sa_handle_t *sa_hdl, sa_attr_type_t *sa_tbl,
+    const char *name, uint64_t *txg)
+{
+	nvlist_t *nvl = NULL;
+	int err = libuzfs_get_nvlist_from_handle(sa_tbl, &nvl, sa_hdl);
+	if (err != 0) {
+		return (err);
+	}
+
+	err = nvlist_remove(nvl, name, DATA_TYPE_BYTE_ARRAY);
+	if (err != 0) {
+		goto out1;
+	}
+
+	uint64_t xattr_sa_size = 0;
+	err = nvlist_size(nvl, &xattr_sa_size, NV_ENCODE_XDR);
+	if (err != 0) {
+		goto out1;
+	}
+
+	char *xattr_sa_data = NULL;
+	if (xattr_sa_size > 0) {
+		xattr_sa_data = vmem_alloc(xattr_sa_size, KM_SLEEP);
+		err = nvlist_pack(nvl, &xattr_sa_data,
+		    &xattr_sa_size, NV_ENCODE_XDR, KM_SLEEP);
+		if (err != 0) {
+			goto out2;
+		}
+	}
+
+
+	dmu_tx_t *tx = dmu_tx_create(sa_hdl->sa_os);
+	dmu_tx_hold_sa(tx, sa_hdl, B_TRUE);
+	if ((err = dmu_tx_assign(tx, TXG_WAIT)) == 0) {
+		if (xattr_sa_size > 0) {
+			ASSERT(xattr_sa_data != NULL);
+			VERIFY0(sa_update(sa_hdl, sa_tbl[UZFS_XATTR],
+			    xattr_sa_data, xattr_sa_size, tx));
+		} else {
+			VERIFY0(sa_remove(sa_hdl, sa_tbl[UZFS_XATTR], tx));
```

Comment:
Will this op change the attr layout? how big the overhead to remove and add an sa attribute frequently?

Reply:
- if remove after an add happens frequently, we need to do the following steps
1. register a new layout(in most circumstances we only need to lookup the new layout in the layout_obj, whose content is cached in memory)
2. change to a new sa header, and arrange attrs according to new header
3. free spill buffer (this will happen when xattr_sa_size is small)
- Currently, xattr_sa_size will never be zero, because the nvlist header takes up some space. so the question is whether we should call sa_remove when the nvlist is empty, zfs didn't do this check